### DISCUSSION TITLE
can one do a diff with a live filesystem

### DISCUSSION THREAD

**Discussion by tshepang** [rustic-rs/rustic#244] [Q&A] [answered]
That is, not comparing snapshots with each other. Would be nice to verify what changed before doing the backup.

**Comment by aawsome** [comment, marked as answer]
This is not completely possibe. There are two posibilites:

1) The `backup` command automatically compares the parent snapshot with the live filesystem. If you want to compare with an arbitrary snapshot, you can manually set the parent using the `--parent` option.

To not backup but still see the changes, you can do a backup dry-run. If you also want to see, which files changed, use log-level debug:
`rustic --log-level debug backup -n ...`

This however has the limitation that files which do not exist in the live system, but exist in the snapshot, are not shown at all..

2) The restore command also has a dry-run option. It knows which files are different, but so far only shows files which only exist in the live system but not in the snapshot. I think the debug output should be a bit more verbose here...

Does one of these options suit your needs or would you like to have more features?

**Reply by tshepang** [reply]
> This however has the limitation that files which do not exist in the live system, but exist in the snapshot, are not shown at all..

would these be deleted files

**Reply by aawsome** [reply]
Yes, exactly. For the `backup` run it is irrelevant what the reason is if a file does not exist. In any case, it will not be contained in the snapshot that is generated by that `backup` run.

I'll make the debug output of `restore` more verbose. `restore` already differentiates these cases:
- file in snapshots equals the file on destination => nothing to do
- file in snapshot exists with same name but different contents on destination => need to restore content
- file exists only in snapshot => need to create and restore content
- files exists only on destination => print a warning or remove if `--delete` is specified.

So all information is already available. It is only not printed out ATM.

**Reply by tshepang** [reply]
These still feel like workarounds... they are good on their own of course, but I feel use-case is important enough to deserve its first-class functionality. I don't know what the UI would be... maybe just `rustic diff` should magically do it, where it compares latest snapshot with live filesystem.

Regardless, am thankful for your help (and for the tool itself).

**Reply by aawsome** [reply]
Yeah, you are right. Improves the `restore` command but feels like a workaround. I initially thought that integrating this in the `diff` command would be really hard. But thinking about it, I just got an idea and now I'm pretty convinced it can be done rather easily.

What would you recon a good syntax to differentiate between snapshots and system path? ATM diff, allows to give two snapshots with optional path(s), e.g.
`rustic diff a12:/home/xxx 12345:/media/bla/home/xxx` or `rustic diff a12 12345:/media/bla`

Directly giving a file system path could cause conflicts as `abc` could be the system path `/abc/` or the snapshot starting with `abc...` So what about `local:<PATH>`, so the command could look like
`rustic diff abc:/home/xxx local:/home/xxx`

Is this a good syntax?

**Reply by tshepang** [reply]
That is fine, but I would remove `local:` prefix, where the unprefixed is magically the raw live filesystem.

As a sidenote, I would also add some convenience to refer to the latest snapshot... `rustic diff latest:/home/xxx /home/xxx`.

**Reply by aawsome** [reply]
Thanks, I also like the syntax without `local:` more, but I don't like the ambiguity the the path can be a hex code and might be the start of a snapshot. But maybe it suffices to document that that path should have a `/` within...

About `latest`: This is more or less easy to add, however then also the filter options should be present for `diff` as `latest` always depends on the subset of snapshots you are considering. But this should be an additional PR then.

EDIT: And I hesitated to add `latest` because this would people make ask about a `latest-1` or even `parent` or whatever you could imagine to do a `rustic diff latest latest-1` or `rustic diff latest parent`.
This is all possible but not high on my prio list... ;-)

**Reply by tshepang** [reply]
> Thanks, I also like the syntax without `local:` more, but I don't like the ambiguity the the path can be a hex code and might be the start of a snapshot. But maybe it suffices to document that that path should have a `/` within...

this, to me, is reason enough to discard the magic... having `local:` there is alright compared to the messy heuristics

> About `latest`: This is more or less easy to add, however then also the filter options should be present for `diff` as `latest` always depends on the subset of snapshots you are considering. But this should be an additional PR then.

can you say more about this *filter options*... I don't understand

> EDIT: And I hesitated to add `latest` because this would people make ask about a `latest-1` or even `parent` or whatever you could imagine to do a `rustic diff latest latest-1` or `rustic diff latest parent`. This is all possible but not high on my prio list... ;-)

I would find it clunky to have to first list snapshots before I could run `diff`, so I would stop at `latest`, as it would be the most commonly used

**Reply by aawsome** [reply]
> can you say more about this filter options... I don't understand

Let's say you backup from more than one host system to a shared repository. Then it might be convenient to use the latest snapshot made from host1 to call `rustic diff latest /local/path` on host1 and the latest snapshot made from host 2 to call `rustic diff latest /local/path` on host2. 

Hence the filter options which are currently available in the `snapshots`, `forget` or `tag` command would be handy to do a `rustic diff --filter-host host1 latest /local/path`. The same applies to the other `--filter-*` options for other backup use cases.

However, I just realized that there are commands like `ls` which already support latest but also miss these filter options.

**Reply by aawsome** [reply]
diff with local files system is now implemented in this PR: #278
Feel free to test it out! 

Note that your "latest" requirement is still open, see #247

**Reply by tshepang** [reply]
looks like it works, but it marks unchanged files with `M` (as if they were modified)

**Reply by aawsome** [reply]
Thanks for testing!  The `M` indicates that the metadata (file permissions, mtime, ctime, size, inode, device_id) is actually different. Do you have a file system which needs `--ignore-inode` or `--ignore-ctime` for `backup` to work with a parent snapshot?

**Reply by aawsome** [reply]
Note that `diff` with a local filesystem currently does not compare the file contents (or even read a local file), but only the metadata.

**Reply by aawsome** [reply]
I merged #278 in its current form into master and opened #282 for additional requirements which might be useful with diff. Feel free to comment in that issue if you have further requirements.

**Reply by tshepang** [reply]
my mount settings are Debian defaults (`ext4 errors=remount-ro`)

**Reply by aawsome** [reply]
I added #365 which enables reading files from the local dir by default and disables comparing metadata by default. 
@tshepang Can you test that PR if this solves your issue with the marks of seemingly unchanged files?

**Reply by tshepang** [reply]
alright, but may take some days before getting there...

**Reply by aawsome** [reply]
#365 is now also in main and beta builds are running... If things go smooth I'll soon release a new version including that PR.

**Reply by aawsome** [reply]
Is included in rustic 0.4.2

**Reply by tshepang** [reply]
> I added #365 which enables reading files from the local dir by default and disables comparing metadata by default. @tshepang Can you test that PR if this solves your issue with the marks of seemingly unchanged files?

Am running 0.4.2, and am not sure how to run a diff against live filesystem. Do I have to know the snapshot in advance?

**Reply by aawsome** [reply]
Yes, you need to know the snapshot you want to compare with the live system. The syntax is `rustic diff <SNAPSHOT>:<PATH-IN-SNAP> <LOCAL-PATH>`, where `<SNAPSHOT>` can also be `latest`.

**Reply by tshepang** [reply]
That is kool, but note that it ignores the (excluded) glob patterns in rustic config.

**Reply by aawsome** [reply]
You are right. You can specify exclude options by the command line.
There is so far no specific config support for `diff` as I'm not sure how to treat it in combination with backup options and backup source options. E.g. should these automatically apply to `diff` or should we have an extra `[diff]` section?

**Reply by tshepang** [reply]
At a minimum, I expected it to ignore files/dirs that I asked to be ignored with glob. Otherwise I'd have to replicate the whole list (with `--iglob-file` I think).

**Comment by aawsome** [comment]
Note that with #305 merged, now `rustic diff latest[:PATH] <PATH>` is possible.

**Reply by tshepang** [reply]
you are amaze!
